From 49a954b08654dd06bab71224a2398a65c2555549 Mon Sep 17 00:00:00 2001
From: Frantisek Sumsal <frantisek@sumsal.cz>
Date: Mon, 15 Apr 2024 15:53:36 +0200
Subject: [PATCH] test: fall back to SYSLOG_IDENTIFIER= matching where necessary

(Partial backport, only for TEST-75-RESOLVED)

Due to systemd/systemd#30886, relying on _SYSTEMD_UNIT= matching might
be unreliable in some cases (with glibc 2.39+) as the journal message
might be missing certain metadata. Since the fix for that issue is too
risky to backport, let's just fall back to SYSLOG_IDENTIFIER= matching
that doesn't seem to have this issue, so we can still run the
"problematic" tests just with some minimal tweaks.

This leaves the skip (from 2d6e263) for the LogFilteringPatterns= stuff
in place, because falling back to SYSLOG_IDENTIFIER= matching doesn't
work there - the output from that tests becomes very weird and I suspect
there's a bug somewhere. However, the same behavior occurs even with the
latest main, so it's not something that's caused by the v255-stable
branch.

v255-only
Partially reverts 2d6e26342997dfc03753e6e6787f950f2fed30df.
---
 test/units/testsuite-75.sh | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/test/units/testsuite-75.sh b/test/units/testsuite-75.sh
index 5423448..7cf279a 100755
--- a/test/units/testsuite-75.sh
+++ b/test/units/testsuite-75.sh
@@ -46,7 +46,8 @@ monitor_check_rr() (
     # displayed. We turn off pipefail for this, since we don't care about the
     # lhs of this pipe expression, we only care about the rhs' result to be
     # clean
-    timeout -v 30s journalctl -u resolvectl-monitor.service --since "$since" -f --full | grep -m1 "$match"
+    # v255-only: match against a syslog tag as well to work around systemd/systemd#30886
+    timeout -v 30s journalctl --since "$since" -f --full _SYSTEMD_UNIT="resolvectl-monitor.service" + SYSLOG_IDENTIFIER="resolvectl-monitor" | grep -m1 "$match"
 )
 
 restart_resolved() {
@@ -251,8 +252,8 @@ resolvectl status
 resolvectl log-level debug
 
 # Start monitoring queries
-systemd-run -u resolvectl-monitor.service -p Type=notify resolvectl monitor
-systemd-run -u resolvectl-monitor-json.service -p Type=notify resolvectl monitor --json=short
+systemd-run -u resolvectl-monitor.service -p SyslogIdentifier=resolvectl-monitor -p Type=notify resolvectl monitor
+systemd-run -u resolvectl-monitor-json.service -p SyslogIdentifier=resolvectl-monitor-json -p Type=notify resolvectl monitor --json=short
 
 # Check if all the zones are valid (zone-check always returns 0, so let's check
 # if it produces any errors/warnings)
@@ -557,10 +558,10 @@ systemctl stop resolvectl-monitor-json.service
 # Issue: https://github.com/systemd/systemd/issues/29580 (part #2)
 #
 # Check for any warnings regarding malformed messages
-(! journalctl -u resolvectl-monitor.service -u reseolvectl-monitor-json.service -p warning --grep malformed)
+(! journalctl -p warning --grep malformed _SYSTEMD_UNIT="resolvectl-monitor-json.service" + SYSLOG_IDENTIFIER="resolvectl-monitor-json")
 # Verify that all queries recorded by `resolvectl monitor --json` produced a valid JSON
 # with expected fields
-journalctl -p info -o cat _SYSTEMD_UNIT="resolvectl-monitor-json.service" | while read -r line; do
+journalctl -p info -o cat _SYSTEMD_UNIT="resolvectl-monitor-json.service" + SYSLOG_IDENTIFIER="resolvectl-monitor-json" | while read -r line; do
     # Check that both "question" and "answer" fields are arrays
     #
     # The expression is slightly more complicated due to the fact that the "answer" field is optional,
